scrapy/core/http2/stream.py

```diff
@@ -59,26 +93,124 @@ def initiate_request(self):
 
             # TODO: Check if scheme can be "http" for HTTP/2 ?
             (":scheme", "https"),
-            (":path", url.path)
+            (":path", url.path),
+
+            # TODO: Make sure 'Content-Type' and 'Content-Encoding' headers
+            #  are sent for request having body
         ]
 
-        self._conn.send_headers(self.stream_id, http2_request_headers)
+        self._conn.send_headers(self.stream_id, http2_request_headers, end_stream=False)
+        self._write_to_transport()
+
+        self.send_data()
+
+    def send_data(self):
+        """Called immediately after the headers are sent. Here we send all the
+         data as part of the request.
+
+         If the content length is 0 initially then we end the stream immediately and
+         wait for response data.
+         """
+
+        # TODO:
+        #  1. Add test for sending very large data
+        #  2. Add test for small data
+        #  3. Both (1) and (2) should be tested for
+        #    3.1 Large number of request
+        #    3.2 Small number of requests
+
+        # Firstly, check what the flow control window is for current stream.
+        window_size = self._conn.local_flow_control_window(stream_id=self.stream_id)
+
+        # Next, check what the maximum frame size is.
+        max_frame_size = self._conn.max_outbound_frame_size
+
+        # We will send no more than the window size or the remaining file size
+        # of data in this call, whichever is smaller.
+        bytes_to_send = min(window_size, self.remaining_content_length)
+
+        # We now need to send a number of data frames.
+        while bytes_to_send > 0:
+            chunk_size = min(bytes_to_send, max_frame_size)
+
+            data_chunk_start = self.content_length - self.remaining_content_length
+            data_chunk = self._request_body[data_chunk_start:data_chunk_start + chunk_size]
 
-    def receive_data(self, data: bytes):
+            self._conn.send_data(self.stream_id, data_chunk, end_stream=False)
+            self._write_to_transport()
+
+            bytes_to_send = max(0, bytes_to_send - chunk_size)
+            self.remaining_content_length = max(0, self.remaining_content_length - chunk_size)
```

Comment:
Given the prior lines:

```
bytes_to_send = min(window_size, self.remaining_content_length)
# …
chunk_size = min(bytes_to_send, max_frame_size)
```

I _think_ it would be safe to assume that the subtraction value will be ≥0.

Reply:
> I think it would be safe to assume that the subtraction value will be ≥0.

It's valid only if `chunk_size` is an integer multiple of both `bytes_to_send` and `self.remaining_content_length`. I removed the `min` though and changed to [this](https://github.com/scrapy/scrapy/pull/4610/commits/01ad8b31ab7fe86fd78a70b09a6dd61a497e0ccb#diff-f9b3d5f8359fe602b12d142e3ed2ace6R143-R148) in 01ad8b31ab7fe86fd78a70b09a6dd61a497e0ccb.